### PR TITLE
Issue534 arbitrary det

### PR DESCRIPTION
**Pull request by adriendaros**
Motivation:
-------------
* see Issue #534 

Main changes:
----------------
* We have created a new method `cryst._get_local_coordinates_of_det()` to compute these translations and rotations.
* Possibility to plot a detector with `cryst.get_detector_approx()` method overlapping exactly the arbitrary one.

Issues:
-------
* Fixes, in devel, Issue #534 

Example:
----------
```
In [1]: import tofu as tf                                                                                                                                                                                                                                                                                                  

In [2]: import inputs_temp.XICS_allshots_C34 as xics                                                                                                                                                                                                                                                                       

In [3]: cryst = tf.load('inputs_temp/TFG_CrystalBragg_ExpWEST_DgXICS_ArXVII_sh00000_Vers1.4.9-315-gc93fa7a3.npz')                                                                                                                                                                                                          
Loaded from:
    /Home/AD265925/project_git/tofu/inputs_temp/TFG_CrystalBragg_ExpWEST_DgXICS_ArXVII_sh00000_Vers1.4.9-315-gc93fa7a3.npz

In [4]: det = dict(np.load('inputs_temp/det37_CTVD_incC4_New.npz', allow_pickle=True))                                                                                                                                                                                                                                     

In [5]: xi, xj = xics._XI, xics._XJ                                                                                                                                                                                                                                                                                        

In [6]: dax = cryst.plot_focal_error_summed(-0.15, 0.15, -0.8, 0.8, xi=xi[::10], xj=xj[::10], err='rel', plot_dets=True, det_ref=det, nsort=0)                                                                                                                                                                             
Computing mean focal error for det (21, 21)/(21, 21)        
/Home/AD265925/project_git/tofu/tofu/geom/_plot_optics.py:1147: UserWarning: The following kwargs were not used by contour: 'linewstyles'
  linewidths=1.,
Parameters of reference detector:
Center position in (x, y, z): ([ 2.3478e+00 -7.7142e+00  4.5000e-03])
Translations (ddist, di, dj): (-0.027332656751099554, -0.0130970408237734, 0.000172915028400119) [m]
Rotations (dtheta, dpsi, tilt): (-0.008141350835571996, -0.6580769831576988, 0.0035090415703947463) [rad]
```


![image](https://user-images.githubusercontent.com/81628304/126797427-4983f6fb-dcb3-431a-abf0-8cc2f5ac89eb.png)

### CODE REVIEW

**Comment by Didou09** [comment]
@adriendaros , so, the tests failed, but the good news is: it is not your fault.

You notice that:
* it fails due another sub-package called openadas
image

* it fails because it did not connect properly to the openadas website (connection https machinchose)
image
 
This is a known problem (identified, but not solved yet)
It happens sometimes, a bit randomly, because te openadas website is overcrowded or somethign like this.

Is such cases: re-run the jobs:

image


And re-run them until they pass

File: tofu/geom/_plot_optics.py [hunk 1]
```diff
@@ -1142,12 +1147,47 @@ def CrystalBragg_plot_focal_error_summed(
...
+                ddist=ddist0,
+                di=di0,
+                dj=dj0,
+                dtheta=dtheta0,
+                dpsi=dpsi0-angle_nout,
+                tilt=tilt0,
+                lamb=lamb,
+                use_non_parallelism=use_non_parallelism,
+                tangent_to_rowland=True,
+            )
```
Comment (Didou09) [inline tofu/geom/_plot_optics.py:1175]: Shorter formulation:

```
        psi0bis = float(psi0)
        if tangent_to_rowland:
            dpsi0bis = dpsi0 - angle_nout

        detector_comp = cryst.get_detector_approx(
            ddist=ddist0,
            di=di0,
            dj=dj0,
            dtheta=dtheta0,
            dpsi=dpsi0bis,
            tilt=tilt0,
            lamb=lamb,
            use_non_parallelism=use_non_parallelism,
            tangent_to_rowland=tangent_to_rowland,
        )
```

Comment (Didou09) [reply tofu/geom/_plot_optics.py:1175]: I am not sure I understand what is the difference between `det_ref` and `detector_comp` ?

Why not use `det_ref` directly?

Was it for debugging ?

File: tofu/geom/_core_optics.py [hunk 2]
```diff
@@ -1938,14 +2067,33 @@ def plot_focal_error_summed(
...
+                        dj=dj0,
+                        # need to re-adapt these rotations to tangent case
+                        dtheta=dtheta0,
+                        dpsi=dpsi0-angle_nout,
+                        tilt=tilt0,
+                        lamb=lamb,
+                        bragg=bragg,
+                        use_non_parallelism=use_non_parallelism,
+                        tangent_to_rowland=True,
+                    )
```
Comment (Didou09) [inline tofu/geom/_core_optics.py:2096]: Shorter formulation:

                psi0bis = float(psi0)
                if tangent_to_rowland:
                    psi0bis = psi0 - angle_nout
                det = self.get_detector_approx(
                    ddist=ddist[ii],
                    di=di[jj],
                    dj=dj0,
                    dtheta=dtheta0,
                    dpsi=dpsi0bis,
                    tilt=tilt0,
                    lamb=lamb,
                    bragg=bragg,
                    use_non_parallelism=use_non_parallelism,
                    tangent_to_rowland=tangent_to_rowland,
                )

**Review by Didou09 — APPROVED** [review]
I approve this PR

It is clear, does the job and does not break the unit tests

However, I have a few questions / comments (see below)

They are not critical (nothing risky), but if possible please correct / answer before we merge (shorter coding formulations possible in 2 places + a question regarding `detector_comp`).

File: tofu/geom/_plot_optics.py (hunk 1 above)
Comment (adriendaros) [reply tofu/geom/_plot_optics.py:1175]: The `_get_local_coordinates_of_det` method provides us the 3 translations (`ddist`, `di`, `dj`) and rotations (`dtheta`, `dpsi`, `tilt`) of the detector of reference `det_ref` which we want to characterize with these variables. 
So `detector_comp` is the detector plotted with the results of this method.
I choose the way to plot both `det_ref` and `detector_comp` to be sure they are overlapping perfectly.
Even if we know that the way to compute these variables is correct, this is an easy way to ensure the method is being correctly its job.
Is this answer good enough ?

File: tofu/geom/_core_optics.py (hunk 2 above)
Comment (adriendaros) [reply tofu/geom/_core_optics.py:2096]: Thanks for that too.

**Comment by Didou09** [comment]
The answers are clear enough, thanks and congratulations, I'm approving and merging

**Pull request merged**